Makefile

```diff
@@ -3,115 +3,102 @@ DIST := skygear-server
 VERSION := $(shell git describe --always)
 GIT_SHA := $(shell git rev-parse HEAD)
 BUILD_DATE := $(shell date -u +"%Y-%m-%dT%H:%M:%SZ")
+TARGETS := gateway auth migrate
 GO_BUILD_LDFLAGS := -ldflags "-X github.com/skygeario/skygear-server/pkg/server/skyversion.version=$(VERSION)"
 GO_TEST_TIMEOUT := 1m30s
 OSARCHS := linux/amd64 linux/386 linux/arm windows/amd64 windows/386 darwin/amd64
 GO_TEST_CPU := 1,4
-GO_TEST_PACKAGE := ./pkg/core/... ./pkg/auth/... ./pkg/gateway/... ./pkg/record/...
+GO_TEST_PACKAGE := ./pkg/core/... ./pkg/auth/... ./pkg/gateway/...
 SHELL := /bin/bash
 
-ifeq (1,${WITH_ZMQ})
-GO_BUILD_TAGS := --tags zmq
-endif
-
 ifeq (1,${GO_TEST_VERBOSE})
 GO_TEST_ARGS_VERBOSE := -v
 endif
 
-DOCKER_COMPOSE_CMD := docker-compose \
-	-f docker-compose.make.yml \
-
-DOCKER_COMPOSE_CMD_TEST := docker-compose \
-	-f docker-compose.test.yml \
-	-p skygear-server-test
-
-ifeq (1,${WITH_DOCKER})
-DOCKER_RUN := ${DOCKER_COMPOSE_CMD} run --rm app
-DOCKER_RUN_DB := ${DOCKER_COMPOSE_CMD_TEST} run --rm db_cmd
-DOCKER_RUN_TEST := ${DOCKER_COMPOSE_CMD_TEST} run --rm app
-GO_TEST_TIMEOUT := 5m
-endif
-
 DOCKER_REGISTRY :=
 DOCKER_ORG_NAME := skygeario
-DOCKER_IMAGE := skygear-server
+DOCKER_IMAGE_AUTH := skygear-auth
+DOCKER_IMAGE_GATEWAY := skygear-gateway
+DOCKER_IMAGE_MIGRATE := skygear-migrate
 DOCKER_TAG := git-$(shell git rev-parse --short HEAD)
 PUSH_DOCKER_TAG := $(VERSION)
-IMAGE_NAME := $(DOCKER_REGISTRY)$(DOCKER_ORG_NAME)/$(DOCKER_IMAGE):$(DOCKER_TAG)
+IMAGE_NAME = $(DOCKER_REGISTRY)$(DOCKER_ORG_NAME)/$(1):$(DOCKER_TAG)
+VERSIONED_IMAGE_NAME = $($(DOCKER_REGISTRY)$(DOCKER_ORG_NAME)/$(1):$(PUSH_DOCKER_TAG))
+AUTH_IMAGE_NAME = $(call IMAGE_NAME,$(DOCKER_IMAGE_AUTH))
+AUTH_VERSIONED_IMAGE_NAME = $(call VERSIONED_IMAGE_NAME,$(DOCKER_IMAGE_AUTH))
+GATEWAY_IMAGE_NAME = $(call IMAGE_NAME,$(DOCKER_IMAGE_GATEWAY))
+GATEWAY_VERSIONED_IMAGE_NAME = $(call VERSIONED_IMAGE_NAME,$(DOCKER_IMAGE_GATEWAY))
+MIGRATE_IMAGE_NAME = $(call IMAGE_NAME,$(DOCKER_IMAGE_MIGRATE))
+MIGRATE_VERSIONED_IMAGE_NAME = $(call VERSIONED_IMAGE_NAME,$(DOCKER_IMAGE_MIGRATE))
 
 GO_BUILD_ARGS := $(GO_BUILD_TAGS) $(GO_BUILD_LDFLAGS)
 GO_TEST_ARGS := $(GO_BUILD_ARGS) -cover -timeout $(GO_TEST_TIMEOUT) $(GO_TEST_ARGS_VERBOSE) -p 1 -cpu $(GO_TEST_CPU)
 
 .PHONY: vendor
 vendor:
-	$(DOCKER_RUN) dep ensure
+	dep ensure
```

Comment:
Didn't review all changes yet, found one problem here.

I think we should keep the `$(DOCKER_RUN)` version for commands, because this is essential for machine without go environment to run these commands.

Same for test case, those commands for preparing docker container for tests should be kept.